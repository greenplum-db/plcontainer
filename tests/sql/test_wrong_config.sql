-- start_ignore
\! plcontainer  runtime-backup -f /tmp/test_backup_cfg_file_wrong_config
-- end_ignore

\! $(pwd)/test_wrong_config.sh 0
select pylog100();

\! $(pwd)/test_wrong_config.sh 1
select pylog100();

\! $(pwd)/test_wrong_config.sh 2
select pylog100();

\! $(pwd)/test_wrong_config.sh 3
select pylog100();

\! $(pwd)/test_wrong_config.sh 4
select pylog100();

\! $(pwd)/test_wrong_config.sh 5
select pylog100();

\! $(pwd)/test_wrong_config.sh 6
select pylog100();

\! $(pwd)/test_wrong_config.sh 7
select pylog100();

\! $(pwd)/test_wrong_config.sh 8
select pylog100();

\! $(pwd)/test_wrong_config.sh 9
select pylog100();

\! $(pwd)/test_wrong_config.sh 10
select pylog100();

\! $(pwd)/test_wrong_config.sh 11
select pylog100();

\! $(pwd)/test_wrong_config.sh 12
select pylog100();

\! $(pwd)/test_wrong_config.sh 13
select pylog100();

\! $(pwd)/test_wrong_config.sh 14
select pylog100();

\! $(pwd)/test_wrong_config.sh 15
select pylog100();

\! $(pwd)/test_wrong_config.sh 16
select pylog100();

\! $(pwd)/test_wrong_config.sh 17
select pylog100();

\! $(pwd)/test_wrong_config.sh 18
select pylog100();

-- start_ignore
 \! plcontainer  runtime-restore -f /tmp/test_backup_cfg_file_wrong_config
-- end_ignore
