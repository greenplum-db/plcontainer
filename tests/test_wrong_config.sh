#!/bin/bash
#------------------------------------------------------------------------------
#
#
# Copyright (c) 2017-Present Pivotal Software, Inc
#
#------------------------------------------------------------------------------
#
# Test bad configuration in plcontainer C code. Called in sql/test_wrong_config.sql

f0 () {
  echo "Test lack of id"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <image>pivotaldata/plcontainer_python:0.0</image>
        <command>./client</command>
    </runtime>
</configuration>
EOF
}

f1 () {
  echo "Test lack of image"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <command>./client</command>
    </runtime>
</configuration>
EOF
}

f2 () {
  echo "Test multiple image"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.0</image>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/> 
        <setting use_container_logging="yes"/>
    </runtime>
</configuration>
EOF
}

f3 () {
  echo "Test multiple id"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <id>plc_python_shared2</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
        <setting use_container_logging="yes"/>
    </runtime>
</configuration>
EOF
}

f4 () {
  echo "Test lack of command element"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.0</image>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
        <setting use_container_logging="yes"/>
    </runtime>
</configuration>
EOF
}

f5 () {
  echo "Test use_container_logging values should only be yes/no"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
        <setting use_container_logging="enable"/>
    </runtime>
</configuration>
EOF
}

f6 () {
  echo "Test duplicate container path."
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients2"/>
        <setting use_container_logging="yes"/>
    </runtime>
</configuration>
EOF
}

f7 () {
  echo "Test wrong use_container_network value."
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
        <setting use_container_network="enable"/>
        <setting use_container_logging="yes"/>
    </runtime>
</configuration>
EOF
}

f8 () {
  echo "Test duplicate id"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
    </runtime>
</configuration>
EOF
}

f9 () {
  echo "Test logs/use_network disable"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting use_container_logging="no"/>
        <setting use_container_network="disable"/>
        <setting memory_mb="512"/>
    </runtime>
</configuration>
EOF
}

f10 () {
  echo "Test wrong setting"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting memory="512"/>
    </runtime>
</configuration>
EOF
}

f11 () {
  echo "Test wrong memory_mb"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting memory_mb="-100"/>
    </runtime>
</configuration>
EOF
}

f12 () {
  echo "Test wrong element"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <images>pivotaldata/plcontainer_python:0.1</images>
        <command>./client</command>
    </runtime>
</configuration>
EOF
}

f13 () {
  echo "Test more than 1 command"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <command>./client2</command>
    </runtime>
</configuration>
EOF
}

f14 () {
  echo "Test 'host' missing in shared_directory"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir"/> 
    </runtime>
</configuration>
EOF
}

f15 () {
  echo "Test 'container' missing in shared_directory"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/> 
    </runtime>
</configuration>
EOF
}

f16 () {
  echo "Test 'access' missing in shared_directory"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/> 
    </runtime>
</configuration>
EOF
}

f17 () {
  echo "Test bad access in shared_directory"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="rx" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/> 
    </runtime>
</configuration>
EOF
}

f18 () {
  echo "Test long runtime id which exceeds the length limit"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared_toooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo_long</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
        <setting use_container_logging="yes"/>
    </runtime>
</configuration>
EOF
}

f19 () {
  echo "Test wrong cpu_share name"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting cpu="100"/>
    </runtime>
</configuration>
EOF
}

f20 () {
  echo "Test wrong cpu_share value"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting cpu_share="-100"/>
    </runtime>
</configuration>
EOF
}

f21 () {
  echo "Test good format (but it still fails since the configuration is not legal)"
  cat >/tmp/bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>not_exist_pivotaldata/plcontainer_python:0.1</image>
        <command>./not_exist_client</command>
        <shared_directory access="rw" container="/clientdir1" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients1"/> 
        <shared_directory access="ro" container="/clientdir2" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients2"/> 
        <setting use_container_logging="no"/>
        <setting use_container_network="yes"/>
        <setting memory_mb="512"/>
        <setting cpu_share="1024"/>
    </runtime>
</configuration>
EOF
}

function _main() {
  local config_id="${1}"
  if [ "$config_id" = "0" ]; then
	  f0
  elif [ "$config_id" = "1" ]; then
	  f1
  elif [ "$config_id" = "2" ]; then
	  f2
  elif [ "$config_id" = "3" ]; then
	  f3
  elif [ "$config_id" = "4" ]; then
	  f4
  elif [ "$config_id" = "5" ]; then
	  f5
  elif [ "$config_id" = "6" ]; then
	  f6
  elif [ "$config_id" = "7" ]; then
	  f7
  elif [ "$config_id" = "8" ]; then
	  f8
  elif [ "$config_id" = "9" ]; then
	  f9
  elif [ "$config_id" = "10" ]; then
	  f10
  elif [ "$config_id" = "11" ]; then
	  f11
  elif [ "$config_id" = "12" ]; then
	  f12
  elif [ "$config_id" = "13" ]; then
	  f13
  elif [ "$config_id" = "14" ]; then
	  f14
  elif [ "$config_id" = "15" ]; then
	  f15
  elif [ "$config_id" = "16" ]; then
	  f16
  elif [ "$config_id" = "17" ]; then
	  f17
  elif [ "$config_id" = "18" ]; then
	  f18
  elif [ "$config_id" = "19" ]; then
	  f19
  elif [ "$config_id" = "20" ]; then
	  f20
  elif [ "$config_id" = "21" ]; then
	  f21
  fi

  if [ -z "$MASTER_DATA_DIRECTORY" ]; then
	  echo "Error: MASTER_DATA_DIRECTORY is not set"
	  exit 1
  else
  	  cp /tmp/bad_xml_file $MASTER_DATA_DIRECTORY/plcontainer_configuration.xml
	  rm -f /tmp/bad_xml_file
  fi
}
_main "$@"
