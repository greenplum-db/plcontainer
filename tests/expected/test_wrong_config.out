-- start_ignore
\! plcontainer  runtime-backup -f /tmp/test_backup_cfg_file_wrong_config
-- end_ignore
\! $(pwd)/test_wrong_config.sh 0
Test lack of id
select pylog100();
ERROR:  plcontainer: tag <id> must be specified in configuartion (plc_configuration.c:131)
\! $(pwd)/test_wrong_config.sh 1
Test lack of image
select pylog100();
ERROR:  plcontainer: Lack of 'image' subelement in a runtime element plc_python_shared (plc_configuration.c:251)
\! $(pwd)/test_wrong_config.sh 2
Test multiple image
select pylog100();
ERROR:  plcontainer: There are more than one 'image' subelement in a runtime element plc_python_shared (plc_configuration.c:248)
\! $(pwd)/test_wrong_config.sh 3
Test multiple id
select pylog100();
ERROR:  plcontainer: tag <id> must be specified only once in configuartion (plc_configuration.c:118)
\! $(pwd)/test_wrong_config.sh 4
Test lack of command element
select pylog100();
ERROR:  plcontainer: Lack of 'command' subelement in a runtime element plc_python_shared (plc_configuration.c:258)
\! $(pwd)/test_wrong_config.sh 5
Test log values should only be enable/disable
select pylog100();
ERROR:  plcontainer: SETTING element <log> only accepted "enable" or"disable" only, current string is yes (plc_configuration.c:194)
\! $(pwd)/test_wrong_config.sh 6
Test duplicate container path.
select pylog100();
ERROR:  plcontainer: Container path cannot be the same in 'shared_directory' element in the runtime plc_python_shared (plc_configuration.c:292)
\! $(pwd)/test_wrong_config.sh 7
Test wrong use_network value.
select pylog100();
ERROR:  plcontainer: SETTING element <use_network> only accepted "yes"|"true" or"no"|"false" only, current string is enable (plc_configuration.c:222)
\! $(pwd)/test_wrong_config.sh 8
Test duplicate id
select pylog100();
ERROR:  plcontainer: tag <id> must be specified only once in configuartion (plc_configuration.c:118)
\! $(pwd)/test_wrong_config.sh 9
Test logs/use_network disable
select pylog100();
ERROR:  plcontainer: SETTING element <use_network> only accepted "yes"|"true" or"no"|"false" only, current string is disable (plc_configuration.c:222)
\! $(pwd)/test_wrong_config.sh 10
Test wrong setting
select pylog100();
ERROR:  plcontainer: Unrecognized setting options, please check the configuration file: plc_python_shared (plc_configuration.c:229)
\! $(pwd)/test_wrong_config.sh 11
Test wrong memory_mb
select pylog100();
ERROR:  plcontainer: container memory size could not less 0, current string is -100 (plc_configuration.c:204)
\! $(pwd)/test_wrong_config.sh 12
Test wrong element
select pylog100();
ERROR:  plcontainer: Unrecognized element 'images' inside of container specification (plc_configuration.c:242)
\! $(pwd)/test_wrong_config.sh 13
Test more than 1 command
select pylog100();
ERROR:  plcontainer: There are more than one 'command' subelement in a runtime element plc_python_shared (plc_configuration.c:255)
\! $(pwd)/test_wrong_config.sh 14
Test 'host' missing in shared_directory
select pylog100();
ERROR:  plcontainer: Configuration tag 'shared_directory' has a mandatory element 'host' that is not found: plc_python_shared (plc_configuration.c:277)
\! $(pwd)/test_wrong_config.sh 15
Test 'container' missing in shared_directory
select pylog100();
ERROR:  plcontainer: Configuration tag 'shared_directory' has a mandatory element 'container' that is not found: plc_python_shared (plc_configuration.c:284)
\! $(pwd)/test_wrong_config.sh 16
Test 'access' missing in shared_directory
select pylog100();
ERROR:  plcontainer: Configuration tag 'shared_directory' has a mandatory element 'access' that is not found: plc_python_shared (plc_configuration.c:300)
\! $(pwd)/test_wrong_config.sh 17
Test bad access in shared_directory
select pylog100();
ERROR:  plcontainer: Directory access mode should be either 'ro' or 'rw', but passed value is 'rx': plc_python_shared (plc_configuration.c:306)
\! $(pwd)/test_wrong_config.sh 18
Test good format (but it still fails since the configuration is not legal)
select pylog100();
ERROR:  plcontainer: Backend create error: Failed to create container (return code: 404). (containers.c:356)
-- start_ignore
 \! plcontainer  runtime-restore -f /tmp/test_backup_cfg_file_wrong_config
20180111:11:27:37:025178 plcontainer:localhost:gpadmin-[INFO]:-Distributing file plcontainer_configuration.xml to all locations...
20180111:11:27:37:025178 plcontainer:localhost:gpadmin-[INFO]:-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
-- end_ignore
